### DISCUSSION TITLE
Web support

### DISCUSSION THREAD

**Discussion by nandorojo** [mmazzarolo/react-native-modal-datetime-picker#599] [Ideas]
## Ask your Question

Are there plans to support web? It would be awesome to have a universal date picker that also works on web.

**Comment by mmazzarolo** [comment]
Hey,
That sounds like a nice idea. I know the community picker supports Windows, but I don't see any plan to support the web anytime soon... We could still build it directly in this project, but at the same time I'm wondering if showing a date/time picker in a modal is a nice UX on the web 🤔

**Comment by PManager1** [comment]
we need the picker working on web ++

**Comment by mmazzarolo** [comment]
> we need the picker working on web ++

PRs, ideas, and suggestions are welcome!

**Comment by nandorojo** [comment]
I agree a modal _might_ not be the best UX on web. But if it is, react-native-web supports modals now. Perhaps a native `input` with a simple date picker could be a good start for web, or some sort of popover using [`react-native-popover-view`](https://github.com/steffeydev/react-native-popover-view#readme). Whatever feels simplest.

**Comment by nandorojo** [comment]
Here is an example of a nice modal picker on web: https://material-ui.com/components/pickers/ Could just fall back to that on web for now?

**Comment by mmazzarolo** [comment]
I'd go for a native solution like this one https://codepen.io/getify/pen/LYZbaGj?editors=1111, which looks native on the web both on desktop and mobile

**Reply by nandorojo** [reply]
The native safari picker is the most beautiful one imo. What's nice about the actually-native picker is that it's up to each browser to implement optimally, and new iOS browsers have been doing so. I wonder if perfect is the enemy of the good when it comes to implementing this library on Web. I'd love to get this discussion closed with a native `<input type="date" />`. 

Would you be open to this @mmazzarolo? 

It's been over a year since I opened this, and I've been using a material date picker on Web in the meantime. But it's just way too bulky and opinionated in style. I'd much rather get a simpler, leaner, browser-native approach.

**Reply by mmazzarolo** [reply]
@nandorojo yeah, I'm open to this 👍 Did you already check if there's any interest in the community-picker to implement the web picker on that side? That's the main thing that makes me wonder if developing it in this repo is the right choice... but I'm ok with starting it here (and eventually move it later on).

**Reply by nandorojo** [reply]
I did check, the maintainers seem set on not implementing it. Kind of a bummer. But I think it should be (somewhat?) simple to add one.

Context: https://github.com/react-native-datetimepicker/datetimepicker/issues/491#issuecomment-968099329

**Reply by mmazzarolo** [reply]
Gotcha. 
Do you wanna start drafting some work or creating a discussion/issue in this repo? 
I don't have much time to implement it myself nowadays, but I can help and provide support 👍

**Reply by nandorojo** [reply]
Yeah I think I could do that when I have some time. I’m deciding if there’s a need to re-implement our picker or not, so the timing is good. I could do it all in a few hours. The only open question is, how does it work with the input itself? Should it display the input on web or not?

**Reply by khevamann** [reply]
I personally am not a fan of letting each browser implement the changes differently. We found that the native implementation behaves quite differently on safari vs chrome. My solution is using [Flatpickr](https://flatpickr.js.org/) which is a lightweight react implementation. Here is an example of how I am using it https://codesandbox.io/s/web-datetime-demo-05nnd. The main reason we chose to go this route is the native datepickers do not provide a lot of key functionality. For example: selecting ranges, disabling input, blocking certain dates. Personally I believe this is the best route as it could match almost all of the native API functionality. Let me know what you all think. I would be happy to create a PR that implements the basic functionality so that it could be messed around with, or sharing my full code that switches between the native `react-native-modal-datetime-picker` on mobile and this implementation on web.

* Another note is safari shows dates in DD-MM-YYYY format and I could not find a good way to customize that which was really confusing some users.

**Reply by nandorojo** [reply]
Could you share your implementation? I’m hoping to use something that is very lean but feature-sufficient. It doesn’t need to be a native solution necessarily, but I don’t want a bulky library ideally. So seeing your usage could help for sure.

**Reply by khevamann** [reply]
@nandorojo I realized it might not give you much more information if I share what I have. Basically I just have a `DateTimePicker.ts` and `DateTimePicker.web.ts` so it automatically chooses the appropriate one. The codesandbox provides the implementation we have, converted from typescript to plain js and removing features we added to handle timezone support. So Basically the way I would implement it is adding the .web.js extension here `https://github.com/mmazzarolo/react-native-modal-datetime-picker/tree/master/src`. I understand your concern about the size, and I guess it's really a tradeoff of how many of the native features to support and how much extra bloat to add. I am not sure how to check the true size of this package, as it has extra styling and support for other libraries that are stripped. I did research it when I initially implemented it and it seemed to be the most lightweight of the other options that have similar feature sets.

**Reply by khevamann** [reply]
Found it.
![image](https://user-images.githubusercontent.com/19052698/146257594-f6d3e7f4-af95-4e76-b138-b2328f9dfd78.png)

**Comment by nandorojo** [comment]
That's true. It doesn't have as much [browser support](https://caniuse.com/input-datetime), but I think the native input makes sense.

**Comment by ltaljaard** [comment]
> I'd go for a native solution like this one https://codepen.io/getify/pen/LYZbaGj?editors=1111, which looks native on the web both on desktop and mobile

The native support is not there for browsers like Safari (the underlying "input" control falls back to text-only on such browsers) so I wouldn't go that route until all browsers have caught up.

I am now using react-datepicker in my web solution which is not terrible, maybe you can just wrap that to add web support for now: [https://www.npmjs.com/package/react-datepicker](url)

It can run in inline mode as well as in portal mode (modal-like), there are a few tricks to get it working properly but not too bad and now that it works it looks quite good compared to the other things i've tried. I can share that code if you're interested.

![Untitled 34](https://user-images.githubusercontent.com/1267342/97720646-308d8480-1ad1-11eb-8f90-d7f33f5cd554.png)

**Comment by nandorojo** [comment]
Yeah that'd be great if you could share, I had some trouble with it when I tried it. Don't love the design but it's simple.

**Comment by ltaljaard** [comment]
I'll just sum it up here. This is for an Expo Web app which will then work on the mobile as well as web.


Add these to your project (in addition to react-native-modal-datetime-picker):
> npm i react-datepicker
> npm i @date-io/date-fns@1.x date-fns
> npm i react-popper @popperjs/core


Import this at the top of your main form:
`import {DateTimePickerModal as MobileDateTimePickerModal} from "react-native-modal-datetime-picker";
import ReactDatePicker from "react-datepicker";
import "ReactDatePicker_CSSHelper";`
The above CSS helper import is to get around the general silliness with regards to import/require and discrepancies between what works on the mobile vs the web app. Add the following two files to your project with the contents as below:

File 1: filename must be "ReactDatePicker_CSSHelper.js":
`import * as css from "react-datepicker/dist/react-datepicker.css";
export default css;`

File 2: filename must be "ReactDatePicker_CSSHelper.native.js":
`export default undefined;`


The state stuff at the top of your main form:
`const initialState = { model: new <your data model here e.g. as defined in schema.graphql or just your own class, your class must include a field called date1, mine is of type AWSDateTime>({ }), ...<whatever else you want to initialize in the initial state>... };`


The state stuff declared at the top inside the function() {...} block:
  `const [formState, updateFormState] = useState(initialState);
  const [isMobileDatePickerVisible_date1, setMobileDatePickerVisibility_date1] = useState(false);`


Add these functions to your main form inside the function() {...} block:
`  function setFormState_model(key, value) {
    updateFormState({...formState, model: {...formState.model, [key]: value } });
  }
const WebDatePicker_date1_CustomTimeInput = ({ date, value, onChange }) => (
    <ReactDatePicker
      selected={formState.model.date1!=undefined ? new Date(formState.model.date1) : new Date()}
      onChange={selectedDati => { setFormState_model("date1", selectedDati.toISOString()) }}
      showTimeSelect
      showTimeSelectOnly
      timeIntervals={1}
      timeCaption="Time"
      dateFormat="HH:mm"
      timeFormat="HH:mm"
    />
  );
  const showMobileDatePicker_date1 = () => {
    setMobileDatePickerVisibility_date1(true);
  };
  const hideMobileDatePicker_date1 = () => {
    setMobileDatePickerVisibility_date1(false);
  };
  const handleMobileDatePickerConfirm_date1 = date => {
    setFormState_model("date1", date.toLocaleString());
    hideMobileDatePicker_date1();
  };
  function _DateTimePicker_date1()
  {
    if(Platform.OS === "ios" || Platform.OS === "android") {
      return(
        <View>
            <MobileDateTimePickerModal
                isVisible={isMobileDatePickerVisible_date1}
                onConfirm={handleMobileDatePickerConfirm_date1}
                onCancel={hideMobileDatePicker_date1}
                mode={"datetime"}
                is24Hour={true}
                date={formState.model.date1!=undefined ? new Date(formState.model.date1) : new Date()}
                value={formState.model.date1}
            />
            <Text style={[globalStyles.detailsLabel]}>Label for date1</Text>
            <Input
                disabled={true}
                onTouchEnd={()=>{showMobileDatePicker_date1(); console.log({touchEnd:"touchEnd"});}}
                placeholder="Placeholder text for date1"
                value={formState.model.date1}
                style={globalStyles.detailsLabel}
            />
        </View>
      );
    } else {
      return(
        <View>
            <Text style={[globalStyles.detailsLabel]}>Label for date1</Text>
            <div>
                <ReactDatePicker
                    portalId="root-portal"
                    selected={ formState.model.date1!=undefined ? new Date(formState.model.date1) : new Date() }
                    onChange={ selectedDati => { setFormState_model("date1", selectedDati.toISOString()); } }
                    peekNextMonth
                    showMonthDropdown
                    showYearDropdown
                    dropdownMode="select"
                    todayButton="Today"
                    timeIntervals={1}
                    customTimeInput={<WebDatePicker_date1_CustomTimeInput/>}
                    showTimeInput
                    dateFormat="yyyy/MM/dd HH:mm"
                />
            </div>
        </View>
      );
    }
  }`

In your main form's return statement:
`  return (
      <View>
          <ScrollView>
                  {_DateTimePicker_date1()}
          </ScrollView>
      </View>
  );`

@mmazzarolo I'm pretty sure the above could easily be pulled into react-native-modal-datetime-picker to provide Expo Web support.

**Comment by mmazzarolo** [comment]
Very happy we're having this discussion 💥 

Let me ask you all a question: why do you feel adding support for the web to this component would be a good idea?  
Since those examples aren't using a modal (it's more of a popover, right?), wouldn't it make more sense if the support was added to the [community datetimepicker instead -- which already supports a similar solution for Windows](https://github.com/react-native-datetimepicker/datetimepicker)?  

To clarify: I'm __not__ against adding support for it here 👍 I'm just wondering if it's the right approach or not, given that  what this modal datetimepicker is doing is just wrapping the community picker and expose it into a unified modal-UI.

**Comment by ltaljaard** [comment]
I hear you, i think maybe it does not belong inside either of these because they are both based on native - and the react-datepicker is not native.

Actually i think it would make most sense to create a new component based on the code i sent earlier which then uses react- native-modal-datetimepicker for mobile and react-datepicker for the web.

The purpose of such a component would simply be to provide a reasonable out-of-the-box component that people can use in an Expo Web / React Native Web app which works from the mobile as well as from the web. Because currently there is nothing even close to decent and cross-browser out there that does both, at least none that i could find. What do you think?

**Comment by mmazzarolo** [comment]
> i think maybe it does not belong inside either of these because they are both based on native

I don't see this as the main problem in all honesty. I think the "native" word in the react-native ecosystem is more scoped to the platform where the code itself run... meaning for the web it will be JS, which is fine.  
My concern here is just that `react-native-modal-datetimepicker` is focused on showing the content on a modal... and the more I think about it, the more I think the approach on the web is leaning toward a popover-like style (which would be totally out-of-scope in this repo).    
I feel that `react-native-datetimepicker` _might_ suit better this use case. That said, I'm pretty sure the maintainers there are quite busy nowadays, so this is something that should be kickstarted by who's interested in it (while still pinging the maintainers beforehand).  
Just my 2 cents.

**Comment by ltaljaard** [comment]
Ok, but i mean the true web native control for picking a date is the <input ... type="date" ...> as per [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date](url) - which is not supported by all modern browsers. So react-datepicker is not the actual native thing. But maybe that doesn't matter that much.

But anyway i just thought i would share my code as i struggled a bit to get it working. I don't have the time right now to do a separate wrapper for it but i think it would work well, so if anyone else wants to run with it please feel free.

I have added a reference to this thread on the react-native-datetimepicker project: [https://github.com/react-native-datetimepicker/datetimepicker/issues/318](url). Maybe they will consider incorporating it there.

**Comment by nandorojo** [comment]
> My concern here is just that react-native-modal-datetimepicker is focused on showing the content on a modal.

I've come to agree with this.

I have tried a number of web date pickers in the past few days. Materials UI's has been the best one to use on web for a number of reasons.

1) great developer experience, easy to use, and the API is very similar to this package
2) the design is great
3) perhaps most importantly, the modal behavior on web feels very consistent with the modal behavior on other devices when used with this library.

I understand the motivation to use a native HTML date picker to stay closer to the browser's api. That said, I think such a picker belongs in `@react-native-community`'s picker. (Turns out the HTML picker has terrible browser support so I maybe still wouldn't use it there, but in principle, that's where it belongs.)

This library is a modal date time picker, and it should have a modal experience on all platforms IMO. The material UI one is very smooth, lightweight, and has a design that's consistent with this modal picker.

**Comment by khevamann** [comment]
Just wanted to put this out there as of Oct 22 Safari now supports the date, time, and date-time picker. So they are fully supported across all browsers except IE. I am using the web-native date pickers and think it's the way to go. However, I would agree that maybe it doesn't belong in this repo.

**Comment by nandorojo** [comment]
Relevant package: https://github.com/web-ridge/react-native-paper-dates

**Comment by mmazzarolo** [comment]
Hey @nigranac ! This is not the right place to discuss your issue :) Please ask on Stackoverflow, thanks!

**Comment by vmtri96** [comment]
Hi all, I want to disable the specified date. I saw all the properties but couldn't see the solution. How I can do that? Please help me. Thank you

**Comment by mmazzarolo** [comment]
https://developer.chrome.com/blog/show-picker/

**Comment by felix-lambert** [comment]
Hello so is it possible to use this package for web?
